src/rpc/rawtransaction_util.cpp

```diff
@@ -280,102 +274,101 @@ CMutableTransaction ConstructTransaction(const UniValue& inputs_in, const UniVal
 
     }
 
-    if (!outputs_is_obj) {
-        // Translate array of key-value pairs into dict
-        UniValue outputs_dict = UniValue(UniValue::VOBJ);
-        for (size_t i = 0; i < outputs.size(); ++i) {
-            const UniValue& output = outputs[i];
-            if (!output.isObject()) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, key-value pair not an object as expected");
-            }
-            if (output.size() != 1) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, key-value pair must contain exactly one key");
-            }
-            outputs_dict.pushKVs(output);
-        }
-        outputs = std::move(outputs_dict);
-    }
     // Keep track of the fee output so we can add it in the very end of the transaction.
     CTxOut fee_out;
 
     // Duplicate checking
     std::set<CTxDestination> destinations;
     bool has_data{false};
 
-    for (const std::string& name_ : outputs.getKeys()) {
+    for (unsigned int i = 0; i < outputs.size(); ++i) {
+        const UniValue& output = outputs[i].get_obj();
+
         // ELEMENTS:
         // Asset defaults to policyAsset
-        CAsset asset(::policyAsset);
-        if (!assets.isNull()) {
-            if (!find_value(assets, name_).isNull()) {
-                asset = CAsset(ParseHashO(assets, name_));
-            }
-        }
+        CTxOut out(::policyAsset, 0, CScript());
 
-        if (name_ == "data") {
-            if (has_data) {
-                throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, duplicate key: data");
-            }
-            has_data = true;
-            std::vector<unsigned char> data = ParseHexV(outputs[name_].getValStr(), "Data");
-
-            CTxOut out(asset, 0, CScript() << OP_RETURN << data);
-            rawTx.vout.push_back(out);
-            if (output_pubkeys_out) {
-                output_pubkeys_out->push_back(CPubKey());
-            }
-        } else if (name_ == "vdata") {
-            // ELEMENTS: support multi-push OP_RETURN
-            UniValue vdata = outputs[name_].get_array();
-            CScript datascript = CScript() << OP_RETURN;
-            for (size_t i = 0; i < vdata.size(); i++) {
-                std::vector<unsigned char> data = ParseHexV(vdata[i].get_str(), "Data");
-                datascript << data;
-            }
+        bool is_fee = false;
+        for (const std::string& name_ : output.getKeys()) {
+            if (name_ == "data") {
+                if (has_data) {
+                    throw JSONRPCError(RPC_INVALID_PARAMETER, "Invalid parameter, duplicate key: data");
+                }
```

Comment:
Might be better to just check that no keys are duplicated rather than special-casing this. (I realize that you are just copying the existing logic here, but the existing logic seems wrong/incomplete.)

Reply:
I would prefer for a followup to do that.